### PR TITLE
Divide SoundSource (for future streaming)

### PR DESCRIPTION
**Pull request by ShFil119**
Not sure about naming convention(especially about `Wrap`).

### CODE REVIEW

**Comment by madebr** [comment]
Is this tested by a unit test?

**Comment by ShFil119** [comment]
Not yet. Music (aka reading normal files) should be easy to do, problem is with sfx.

File: rwengine/src/audio/SoundSource.cpp [hunk 1]
```diff
@@ -24,29 +23,91 @@ extern "C" {
...
+    buf_size = std::min(buf_size, static_cast<int>(input->size));
+    /* copy internal data to buf */
+    memcpy(buf, input->ptr, buf_size);
+    input->ptr += buf_size;
+    input->size -= buf_size;
+    return buf_size;
+}
+}  // namespace
+
+bool SoundSource::prepareFormatConstextSfx(LoaderSDT& sdt, size_t index,
```
Comment (danhedron) [inline rwengine/src/audio/SoundSource.cpp]: *`Context`

File: rwengine/src/audio/SoundSource.cpp [hunk 2]
```diff
@@ -24,29 +23,91 @@ extern "C" {
...
+    /// Cast pointer, in order to match required layout for ffmpeg
+    input.ptr = reinterpret_cast<uint8_t*>(raw_sound.get());
+
+    /// Finally prepare our "custom" format context
+    avioContext = avio_alloc_context(ioBuffer, ioBufferSize, 0, &input,
+                                     &read_packet, nullptr, nullptr);
+    formatContext = avformat_alloc_context();
+    formatContext->pb = avioContext;
+
+    if (avformat_open_input(&formatContext, "nothint", nullptr, nullptr) != 0) {
```
Comment (danhedron) [inline rwengine/src/audio/SoundSource.cpp]:
```suggestion
    if (avformat_open_input(&formatContext, "SDT", nullptr, nullptr) != 0) {
```

File: tests/test_AudioLoading.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#include "test_Globals.hpp"
+
+#include <engine/GameWorld.hpp>
+
+BOOST_AUTO_TEST_SUITE(AudioLoadingTests)
+
+#if RW_TEST_WITH_DATA
```
Comment (danhedron) [inline tests/test_AudioLoading.cpp]: This needs to use the test predicate from #693

File: rwengine/src/audio/SoundSource.cpp (hunk 1 above)
Comment (ShFil119) [reply rwengine/src/audio/SoundSource.cpp]: done

File: rwengine/src/audio/SoundSource.cpp (hunk 2 above)
Comment (ShFil119) [reply rwengine/src/audio/SoundSource.cpp]: done

File: tests/test_AudioLoading.cpp (hunk 3 above)
Comment (ShFil119) [reply tests/test_AudioLoading.cpp]: done

File: cmake/modules/FindFFmpeg.cmake [hunk 4]
```diff
@@ -84,10 +84,18 @@ find_package_handle_standard_args(FFmpeg
     )
 set(FFMPEG_FOUND "${FFmpeg_FOUND}")
 
-if(FFmpeg_FOUND)
-    add_library(ffmpeg::ffmpeg INTERFACE IMPORTED)
-    set_property(TARGET ffmpeg::ffmpeg
-        PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${FFMPEG_INCLUDE_DIR})
-    set_property(TARGET ffmpeg::ffmpeg
-        PROPERTY INTERFACE_LINK_LIBRARIES ${FFMPEG_LIBRARIES})
+find_package(Threads REQUIRED)
```
Comment (ShFil119) [inline cmake/modules/FindFFmpeg.cmake]: @madebr, is it ok?

File: tests/test_AudioLoading.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+BOOST_AUTO_TEST_SUITE(AudioLoadingTests, DATA_TEST_PREDICATE)
+
+// @todo Shfil119 implement
+// This test requires assets
+BOOST_AUTO_TEST_CASE(testBufferIsPlaying) {
+    SoundManager manager{Global::get().e};
+
+    auto audioPath =
+        Global::get().e->data->index.findFilePath("audio/A1_a.wav");
```
Comment (ShFil119) [inline tests/test_AudioLoading.cpp]: Can we legally use assets from demo of gta2?

**Comment by ShFil119** [comment]
I'm merging, if you think there's something to change, then I'll try to cover it in next pull.

**Pull request merged**